conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureContexts.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.undertow.runtime;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.palantir.conjure.java.undertow.lib.Contexts;
+import com.palantir.conjure.java.undertow.lib.Endpoint;
+import com.palantir.conjure.java.undertow.lib.RequestContext;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.util.HeaderValues;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+enum ConjureContexts implements Contexts {
+    INSTANCE;
+
+    @Override
+    public RequestContext createContext(HttpServerExchange exchange, Endpoint _endpoint) {
+        return new ConjureServerRequestContext(exchange);
+    }
+
+    private static final class ConjureServerRequestContext implements RequestContext {
+
+        private final HttpServerExchange exchange;
+
+        private ImmutableListMultimap<String, String> cachedQueryParams;
```

Comment:
Do we not need things to be threadsafe in these contexts? I know it's just a cache, so probs doesn't matter if you instantiate it twice

Reply:
Right, this is lazily initialized. If it's created multiple times, that's fine.

In general we expect requests to be single threaded, that is to say zero to one threads "own" the request at any given point in time, although they may have dispatched N threads to do work, those threads aren't interacting directly with the request.